src/System.Management.Automation/engine/regex.cs

```diff
@@ -171,7 +159,33 @@ private void Init()
                         : StringComparison.CurrentCulture;
                 }
 
-                _isMatch = str => string.Equals(str, Pattern, stringComparison);
+                return stringComparison;
+            }
+
+            if (_isMatch != null)
+            {
+                return;
+            }
+
+            if (Pattern.Length == 1 && Pattern[0] == '*')
+            {
+                _isMatch = s_matchAll;
+                return;
+            }
+
+            int index = Pattern.IndexOfAny(s_specialChars);
+            if (index == -1)
+            {
+                // No special characters present in the pattern, so we can just do a string comparison.
+                _isMatch = str => string.Equals(str, Pattern, GetStringComparison());
+                return;
+            }
+
+            if (index == Pattern.Length - 1 && Pattern[index] == '*')
+            {
+                // No special characters present in the pattern before last position and last character is asterisk.
+                var patternWithoutAsterisk = Pattern.Substring(0, Pattern.Length - 1);
```

Comment:
Minor: replace `Pattern.Length - 1` with `index`.

Reply:
Can you avoid the string copy?

Reply:
We could do:
```c#
                _isMatch = str => str.AsSpan().StartsWith(Pattern.AsSpan().Slice(0, index), GetStringComparison());
```
In the code it seems we do more work - convert to span both string and slice Pattern string in every call. Also we can not pass the Span (ref local) to the anonymous method.
```c#
                var patternWithoutAsterisk = Pattern.AsMemory().Slice(0, index);
                _isMatch = str => str.AsSpan().StartsWith(patternWithoutAsterisk.Span, GetStringComparison());
```
I don't know what is better.

Reply:
I measured both cases and second proposal is nice:

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.557 (1809/October2018Update/Redstone5)
Intel Core i3-4150 CPU 3.50GHz (Haswell), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=3.0.100-preview5-011568
  [Host]     : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT


```
|       Method | pattern |          text |     Mean |    Error |   StdDev |   Median | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |-------- |-------------- |---------:|---------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
|     **TestOrig** |      **a*** | **abcdefghijklm** | **169.7 ns** | **3.417 ns** | **7.501 ns** | **166.6 ns** |  **1.00** |    **0.00** | **0.1018** |     **-** |     **-** |     **160 B** |
| TestAsMemory |      a* | abcdefghijklm | 163.1 ns | 3.765 ns | 4.336 ns | 161.8 ns |  0.94 |    0.04 | 0.0918 |     - |     - |     144 B |
|              |         |               |          |          |          |          |       |         |        |       |       |           |
|     **TestOrig** |    **abc*** | **abcdefghijklm** | **196.6 ns** | **3.546 ns** | **2.961 ns** | **195.8 ns** |  **1.00** |    **0.00** | **0.1070** |     **-** |     **-** |     **168 B** |
| TestAsMemory |    abc* | abcdefghijklm | 194.1 ns | 4.193 ns | 7.453 ns | 191.2 ns |  1.00 |    0.04 | 0.0918 |     - |     - |     144 B |
|              |         |               |          |          |          |          |       |         |        |       |       |           |
|     **TestOrig** |  **abcde*** | **abcdefghijklm** | **226.8 ns** | **2.748 ns** | **2.294 ns** | **226.6 ns** |  **1.00** |    **0.00** | **0.1070** |     **-** |     **-** |     **168 B** |
| TestAsMemory |  abcde* | abcdefghijklm | 224.0 ns | 5.349 ns | 7.141 ns | 221.1 ns |  0.99 |    0.04 | 0.0918 |     - |     - |     144 B |

Reply:
Thanks @iSazonov for measuring the diffs